### PR TITLE
fix shared replication log by using sequential writes

### LINKED ISSUE #476: write mutation log bug
Both weijie and I meet the same coredump when kill replica-server repeatedly:

```
D09:23:10.588 (1468920190588449659 4ab4) replica.default0.00004a8600010001: replica.stub:204:initialize(): start to replay shared log
D09:23:10.588 (1468920190588922034 4ab4) replica.default0.00004a8600010001: mutation_log:429:open(): open shared log /home/work/ssd1/pegasus/c3tst-v4/replica/slog/log.32.785532070 succeed, start_offset = 785532070, end_offset = 819086506, size = 33554436
D09:23:10.589 (1468920190589296165 4ab4) replica.default0.00004a8600010001: mutation_log:429:open(): open shared log /home/work/ssd1/pegasus/c3tst-v4/replica/slog/log.33.819086506 succeed, start_offset = 819086506, end_offset = 852641112, size = 33554606
D09:23:10.591 (1468920190591251145 4ab4) replica.default0.00004a8600010001: mutation_log:429:open(): open shared log /home/work/ssd1/pegasus/c3tst-v4/replica/slog/log.34.852641112 succeed, start_offset = 852641112, end_offset = 880403431, size = 27762319
D09:23:10.591 (1468920190591542694 4ab4) replica.default0.00004a8600010001: mutation_log:710:replay(): start to replay mutation log /home/work/ssd1/pegasus/c3tst-v4/replica/slog/log.32.785532070, offset = [785532070, 819086506), size = 33554436
E09:23:11.119 (1468920191119837061 4ab4) replica.default0.00004a8600010001: mutation_log:1818:read_next_log_block(): crc checking failed
D09:23:11.119 (1468920191119878881 4ab4) replica.default0.00004a8600010001: mutation_log:766:replay(): finish to replay mutation log /home/work/ssd1/pegasus/c3tst-v4/replica/slog/log.32.785532070, err = ERR_INVALID_DATA
E09:23:11.119 (1468920191119910844 4ab4) replica.default0.00004a8600010001: mutation_log:883:replay(): replay mutation log failed: ERR_INVALID_DATA
E09:23:11.119 (1468920191119976007 4ab4) replica.default0.00004a8600010001: replica.stub:244:initialize(): replay shared log failed, err = ERR_INVALID_DATA, time_used = 531 ms, clear all logs ...
W09:23:11.121 (1468920191121356995 4ab4) replica.default0.00004a8600010001: dsn.file_utils:640:rename_path(): rename from '/home/work/ssd5/pegasus/c3tst-v4/replica/reps/1.5.rrdb' to '/home/work/ssd5/pegasus/c3tst-v4/replica/reps/1.5.rrdb.err' failed, err = Directory not empty
F09:23:11.121 (1468920191121372573 4ab4) replica.default0.00004a8600010001: /home/work/pegasus/rDSN/src/dist/replication/lib/replica_stub.cpp:261:initialize(): assertion expression: false
F09:23:11.121 (1468920191121381780 4ab4) replica.default0.00004a8600010001: /home/work/pegasus/rDSN/src/dist/replication/lib/replica_stub.cpp:261:initialize(): we cannot recover from the above error, exit ...
```

We doubt  it is caused by unprotected concurrent write of mutation log.
@imzhenyu , would you please look into it?

### PR DESCRIPTION
**Pull request by imzhenyu**
fix shared replication log by using sequential writes instead of concurrent writes; the latter may lead to log file holes. This pr aims at fix #476

### CODE REVIEW

**Comment by imzhenyu** [comment]
@qinzuoyan @shengofsun please help review.

File: src/dist/replication/lib/mutation_log.cpp [hunk 1]
```diff
@@ -52,72 +52,190 @@ using namespace ::dsn::service;
...
+                    write_pending_mutations(true);
+                }
+                else
+                {
+                    _slock.unlock();
+                }
+            }
         },
-        gpid_to_hash(mu->data.header.pid)
+        0
```
Comment (shengofsun) [inline src/dist/replication/lib/mutation_log.cpp:214]: The same question which I've mentioned before: will hash value 0 make the task work 0 have more burden than other workers, as "LPC_WRITE_REPLICATION_LOG_SHARED" is dispatched to thread_pool_replication, and this pool is partitioned?

Comment (imzhenyu) [inline src/dist/replication/lib/mutation_log.cpp:214]: ah, `LPC_WRITE_REPLICATION_LOG_SHARED` needs to be dispatched to the non-replication thread pool now as we now further enqueue the real callbacks in it. Or, if we can make `flush` unnecessary (e.g., using direct io or something), we can `allow_inline` this task to reduce the latency. Let's fix the configurations together with when the flush thing is fixed.

**Comment by qinzuoyan** [comment]
looks good to me, you can merge it now.

**Comment by qinzuoyan** [comment]
@imzhenyu , because I forked branch [pegasus](https://github.com/imzhenyu/rDSN/tree/pegasus) from early point of branch master, so 'pegasus' missed this pull request. Because there are several refactor commits between 'master' and 'pegasus', so we can not apply the fix just by merge, but need to migrate this pull request to 'pegasus'.... That's boring work.

**Comment by qinzuoyan** [comment]
ok, I find I can use cherry-pick to apply this fix

**Comment by qinzuoyan** [comment]
@imzhenyu , the next thing is to fix LPC_WRITE_REPLICATION_LOG_SHARED dispatch thread problem?

**Comment by imzhenyu** [comment]
I guess @shengofsun is working on fixing the flush issue? Once that is fixed, this one is gone as well by simply doing some configuration, as discussed above.

**Comment by qinzuoyan** [comment]
@imzhenyu , this fix (even without O_SYNC) will cause performance degradation for about 100x.
Original set operation is about 1ms, but now set operation takes more than 100ms.

**Comment by qinzuoyan** [comment]
I found the reason, because you invoked lf->flush() when write shared log completed, which cause obvious performance degradation. 
In old version, "lf->flush()" is controlled by "log_shared_force_flush" config, and set as false by default.

**Comment by imzhenyu** [comment]
@qinzuoyan yes, it is a temporal arrangement and it should be gone after the flush issue is fixed.

**Comment by qinzuoyan** [comment]
And there is still bug when using weak_ptr "_issued_write" to determine if is writing, which may cause assert failure in mutation_log_shared::write_pending_mutations() at **dassert(_issued_write.expired(), "");**

The reason is: under undefined thread scheduling, the 'blk' variable in mutation_log_shared::write_pending_mutations() maybe destroyed after write callback invoked.

This bug is fixed by https://github.com/imzhenyu/rDSN/commit/a8148566035059b7fc314314c05c12c635bfe152 before, but brought in again.

**Comment by qinzuoyan** [comment]
That's why I think frequent refactor is **dangerous**, because you may bring in **old** or **new** bugs. We spent much time on fixing bugs, but after refactor, we may need to spend time to fix them again.

**Comment by imzhenyu** [comment]
@qinzuoyan I totally understand your concern, but unfortunately it is always the case as long as we want to advance the system. A better way for you may be to stick to a stable version first for a while, and switch to the next stable one later to catch up. 

As for this particular bug, I think a much simpler fix is to move statement `block == nullptr` before the for loop above, compared to the solution in https://github.com/imzhenyu/rDSN/commit/a8148566035059b7fc314314c05c12c635bfe152.

**Comment by imzhenyu** [comment]
@qinzuoyan btw, even the ``init_prepare` in those callback do not wrte due to that the block is not expired, the latter try after enqueuer will start the write for the pending mutations. I don't quite get it why the situation you describe may happen?

**Comment by qinzuoyan** [comment]
```
(gdb) bt
#0  0x00007f3c7aea6cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f3c7aeaa0d8 in __GI_abort () at abort.c:89
#2  0x00007f3c7bf456b2 in dsn_coredump () at /home/mi/git.test/rDSN.imzhenyu/src/core/core/service_api_c.cpp:307
#3  0x00007f3c79e041e8 in dsn::replication::mutation_log_shared::write_pending_mutations (this=0x7f3bbc002ea0, release_lock=true)
    at /home/mi/git.test/rDSN.imzhenyu/src/dist/replication/lib/mutation_log.cpp:140
#4  0x00007f3c79e03ecb in dsn::replication::mutation_log_shared::__lambda11::operator() (__closure=0x7f3b9c7daff0, err=..., sz=180)
    at /home/mi/git.test/rDSN.imzhenyu/src/dist/replication/lib/mutation_log.cpp:206
```

The reason is: under undefined thread scheduling, the local variable 'blk' in mutation_log_shared::write_pending_mutations() maybe destroyed after line 206 is executed.

**Comment by imzhenyu** [comment]
This is weird. We set `block == nullptr` on line 203, meaning that before line 203 `_issuring_write.expired() == false` and line 203 will make it expire. Since line 203 and later checking are in the same lock, the above assertion failure should not happen. 

Anyway, I just push a pr at #478, please take a  look.

**Comment by qinzuoyan** [comment]
This line:

```
auto blk = std::move(_pending_write);
```

the 'blk' may be holding the shared_ptr.

**Comment by imzhenyu** [comment]
Yes, that why we can `expire` the issued write by using `block = nullptr;` - still don't get it why it leads to crash :-(

**Comment by qinzuoyan** [comment]
You can imagine the thread is suspended at mutation_log.cpp:160, and another thread execute mutation_log.cpp:206.

I just push a pr at #479 .

**Comment by imzhenyu** [comment]
This seems impossible for me as a second thread cannot hit this line when the first thread does not expire issued_write. Can you elaborate in more details about this bug?

**Comment by qinzuoyan** [comment]
I know why:

```
            // start to write if possible
            if (err == ERR_OK)
            {
                _slock.lock();

                if (!_is_writing && _pending_write)
                {
                    write_pending_mutations(true);
                }
                else
                {
                    _slock.unlock();
                }
            }
```

pay attention to the new code: **if (!_is_writing && _pending_write)**
comparing to the old code: **if (_pending_write)**
this is the real reason of coredump.

But what I am talking above is another case (even with small probability):
- thread A invoke write_pending_mutations() to start write
- it store shared_ptr in 'blk' by **auto blk = std::move(_pending_write);**
- then it call **_issued_write_task = pr.first->commit_log_block()** to start write, but at this point, thread A is suspended (but blk is still holding shared_ptr)
- after some time, write done, in callback thread B, another write_pending_mutations() is invoked, but _issued_write is not expired, because shared_ptr is hold by 'blk' in thread A.
- then core at: **dassert(_issued_write.expired(), "");**

**Comment by imzhenyu** [comment]
Can we call? I'm still not convinced:)

**Comment by imzhenyu** [comment]
Thanks for the call. So the reason is that the local `blk` adds an extra reference to the block which makes it impossible for `block == nullptr` to expire `_issued_write`. Nice finding. Let's fix it accordingly using #479

**Pull request merged**